lib/Target/SubstraitPB/Export.cpp

```diff
@@ -757,6 +758,86 @@ FailureOr<std::unique_ptr<Rel>> SubstraitExporter::exportOperation(JoinOp op) {
   return rel;
 }
 
+FailureOr<std::unique_ptr<Rel>>
+SubstraitExporter::exportOperation(HashJoinOp op) {
+  auto relCommon = std::make_unique<RelCommon>();
+  auto direct = std::make_unique<RelCommon::Direct>();
+  relCommon->set_allocated_direct(direct.release());
+
+  auto leftOp =
+      llvm::dyn_cast_if_present<RelOpInterface>(op.getLeft().getDefiningOp());
+  if (!leftOp)
+    return op->emitOpError(
+        "left input was not produced by Substrait relation op");
+
+  FailureOr<std::unique_ptr<Rel>> leftRel = exportOperation(leftOp);
+  if (failed(leftRel))
+    return failure();
+
+  auto rightOp =
+      llvm::dyn_cast_if_present<RelOpInterface>(op.getRight().getDefiningOp());
+  if (!rightOp)
+    return op->emitOpError(
+        "right input was not produced by Substrait relation op");
+
+  FailureOr<std::unique_ptr<Rel>> rightRel = exportOperation(rightOp);
+  if (failed(rightRel))
+    return failure();
+
+  auto hashJoinRel = std::make_unique<HashJoinRel>();
+  hashJoinRel->set_allocated_common(relCommon.release());
+  hashJoinRel->set_allocated_left(leftRel->release());
+  hashJoinRel->set_allocated_right(rightRel->release());
+  hashJoinRel->set_type(static_cast<HashJoinRel::JoinType>(op.getJoinType()));
+
+  if (op.getCondition().empty()) {
+    return op->emitOpError("missing join condition");
+  }
+
+  Block &conditionBlock = op.getCondition().front();
+  Operation *terminator = conditionBlock.getTerminator();
+  Value conditionValue = cast<YieldOp>(terminator).getValue()[0];
+
+  auto compareOp =
+      dyn_cast_or_null<KeyComparisonOp>(conditionValue.getDefiningOp());
+  if (!compareOp) {
+    return op->emitOpError("join condition must be a KeyComparisonOp");
+  }
```

Comment:
This could be removed, since you're already checking for this in your verifier. 
nit: remove braces.